### ISSUE TITLE
Investigate using OverlayFS for nixos containers

### ISSUE THREAD

**Issue by spacekitteh** [NixOS/nixpkgs#21183]
Now that OverlayFS [supports SELinux](https://www.phoronix.com/scan.php?page=news_item&px=Linux-4.9-Security), it is much more suitable for linux containers. Thus, it should be investigated.

(Sidenote: how do nixos-containers and LXC compare?)

**Comment by joachifm** [comment]
NixOS containers are systemd-nspawn+bind mounts. My impression is that overlayfs is commonly used to run container instances off of a (shared) pristine OS runtime (to maximize sharing & having a sane way to do upgrades), but that's not too relevant for us, I think.

**Comment by fpletz** [comment]
I've some changes in my stash which bind-mount only the needed store paths into a container. I didn't push it yet because imperative containers are broken with this code and supporting both would require some refactoring.

It's remarkable though that I didn't notice big differences in performance while (unscientifically) running the container tests on my laptop. We also should check how many mounts are possible, both per namespace and globally in the kernel.

To protect the nix store of the host we only have two solutions: Either via bind mounts or by copying the system closure into a new nix store. Currently, nixos-containers even use the nix-daemon of the host which while being handy is a major security risk because it's running as root.

Does it make sense to support both approaches? For imperative containers, we would need support for copying the system closure anyway if we want to separate the store of the host and the container.

**Comment by Mic92** [comment]
@fpletz yes it makes sense to have both options: copy vs bind mounts - we did something for to build rkt images: https://github.com/Mic92/nix2aci. How did you perform the mounts? My idea was make make an shared bind mount of an arbitrary directory (let's called it `/mnt/containerstore`) in the host to container to `/nix/store`, so that mount events are propagated. In this directory every packages, which is used by container can be bind mounted on the host to `/mnt/containerstore`. This approach has the advantage that containers can be updated without performing a restart.

**Comment by grahamc** [comment]
FWIW OverlayFS is not POSIX-complete: (search for POSIX here:) https://docs.docker.com/engine/userguide/storagedriver/overlayfs-driver/

**Comment by jpotier** [comment]
One interesting use case would be the deployment of web-based applications such as wordpress, dokuwiki, owncloud, nextcloud, etc. We can package a fixed version of their source and mount them with overlayFS for each instance we want to spawn. That should make upgrade easy, as well as giving a sane default to every containers while not including `php-fpm`, `lighttpd`, database drivers, etc. which are still managed by nix.

**Comment by mmahut** [comment]
Any news on this issue?

**Comment by spacekitteh** [comment]
Still important
